packages/expo-battery/android/src/main/java/expo/modules/battery/BatteryModule.kt

```diff
@@ -1,103 +1,155 @@
-  }
+    Constants("isSupported" to true)
 
-  override fun onCreate(moduleRegistry: ModuleRegistry) {
-    val eventEmitter = moduleRegistry.getModule(EventEmitter::class.java)
-    context.registerReceiver(BatteryStateReceiver(eventEmitter), IntentFilter(Intent.ACTION_BATTERY_CHANGED))
-    context.registerReceiver(PowerSaverReceiver(eventEmitter), IntentFilter("android.os.action.POWER_SAVE_MODE_CHANGED"))
-    val ifilter = IntentFilter().apply {
-      addAction(Intent.ACTION_BATTERY_LOW)
-      addAction(Intent.ACTION_BATTERY_OKAY)
+    Events(
+      BATTERY_LEVEL_EVENT_NAME,
+      BATTERY_CHARGED_EVENT_NAME,
+      POWER_MODE_EVENT_NAME
+    )
+
+    OnCreate {
+      registerBroadcastReceivers(context)
     }
-    context.registerReceiver(BatteryLevelReceiver(eventEmitter), ifilter)
-  }
 
-  @ExpoMethod
-  fun getBatteryLevelAsync(promise: Promise) {
-    val batteryIntent: Intent? = context.applicationContext.registerReceiver(
-      null,
-      IntentFilter(Intent.ACTION_BATTERY_CHANGED)
-    )
-    if (batteryIntent == null) {
-      promise.resolve(-1)
-      return
+    OnDestroy {
+      unregisterBroadcastReceivers(context)
+    }
+
+    OnActivityEntersForeground {
+      registerBroadcastReceivers(context)
+    }
+
+    OnActivityEntersBackground {
+      unregisterBroadcastReceivers(context)
+    }
+
+    AsyncFunction("getBatteryLevelAsync") {
+      val batteryIntent: Intent = context.applicationContext.registerReceiver(
+        null,
+        IntentFilter(Intent.ACTION_BATTERY_CHANGED)
+      ) ?: return@AsyncFunction -1
+
+      val level = batteryIntent.getIntExtra(BatteryManager.EXTRA_LEVEL, -1)
+      val scale = batteryIntent.getIntExtra(BatteryManager.EXTRA_SCALE, -1)
+      val batteryLevel = if (level != -1 && scale != -1) {
+        level / scale.toFloat()
+      } else {
+        -1f
+      }
+
+      return@AsyncFunction batteryLevel
+    }
+
+    AsyncFunction("getBatteryStateAsync") {
+      val batteryIntent: Intent = context.applicationContext.registerReceiver(
+        null,
+        IntentFilter(Intent.ACTION_BATTERY_CHANGED)
+      ) ?: return@AsyncFunction BatteryState.UNKNOWN.value
+
+      val status = batteryIntent.getIntExtra(BatteryManager.EXTRA_STATUS, -1)
+      return@AsyncFunction batteryStatusNativeToJS(status).value
+    }
+
+    AsyncFunction("isLowPowerModeEnabledAsync") {
+      isLowPowerModeEnabled
     }
-    val level = batteryIntent.getIntExtra(BatteryManager.EXTRA_LEVEL, -1)
-    val scale = batteryIntent.getIntExtra(BatteryManager.EXTRA_SCALE, -1)
-    val batteryLevel = if (level != -1 && scale != -1) {
-      level / scale.toFloat()
-    } else {
-      -1f
+
+    AsyncFunction("isBatteryOptimizationEnabledAsync") {
+      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+        val packageName: String = context.applicationContext.packageName
+        val powerManager = context.applicationContext.getSystemService(Context.POWER_SERVICE) as PowerManager?
```

Comment:
```suggestion
        val powerManager = context.applicationContext.getSystemService(Context.POWER_SERVICE) as? PowerManager
```
Not a big deal but probably better to use a safe cast here